### PR TITLE
tide: time out when writing to GCS

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Currently it's possible to deadlock on writing to GCS in the tide
history handler. This change adds a timeout to the flush and some
re-tries in case we hit the timeout.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @cjwagner

### CODE REVIEW

File: prow/tide/history/history.go [hunk 1]
```diff
@@ -83,22 +83,53 @@ func readHistory(maxRecordsPerKey int, opener io.Opener, path string) (map[strin
...
+	// data sets can finish in about 500ms; a timeout of 30s should not evict
+	// well-behaved writes
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	// we need to keep track of whether the context timed out or not
+	var timedOut bool
+	go func(done <-chan struct{}) {
+		<-done
+		timedOut = true
```
Comment (stevekuznetsov) [inline prow/tide/history/history.go]: I think this may be racy?

Comment (stevekuznetsov) [reply prow/tide/history/history.go]: Would love ideas of how else to do this.

Comment (fejta) [reply prow/tide/history/history.go]: I don't think we need all this extra complexity -- go routines, a whole second method.

Let's just make `writeHistory()` use a 30s timeout. If it fails we'll try again in a few minutes.

**Comment by stevekuznetsov** [comment]
@cjwagner @fejta updated to do the simple thing

**Pull request merged**